### PR TITLE
Check request labels in Aggregator example

### PR DESCRIPTION
**Pull request by ipetr0v**
This change:
- Adds Router Node to Aggregator example
- Checks client request labels in Aggregator example

Fixes https://github.com/project-oak/oak/issues/1634

### CODE REVIEW

**Comment by tiziano88** [comment]
Not sure I follow this PR, I thought all incoming requests are expected to be with the same label, so we can probably just check in the current "main" node, basically here:

https://github.com/project-oak/oak/blob/7dca63b78edc241e19807191cb2a42396a51dc62/examples/aggregator/module/rust/src/lib.rs#L246-L251

**Comment by ipetr0v** [comment]
I just thought to refactor it also, since we start using `Router` pattern everywhere - so it all will look similar.

**Comment by tiziano88** [comment]
Got it, thanks, so the previous main node is now the router node (more or less), right?

BTW I think https://github.com/project-oak/oak/pull/1647 (if submitted before this) would simplify things a bit.

**Comment by ipetr0v** [comment]
> Got it, thanks, so the previous main node is now the router node (more or less), right?
Correct

> BTW I think #1647 (if submitted before this) would simplify things a bit.
Sure, I'll rebase on it after it's merged

**Comment by tiziano88** [comment]
FYI #1647 is merged. Also I think #1653 might help with this PR, feel free to take a look at that and comment there too.

File: examples/aggregator/module/rust/src/lib.rs
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs:72]: Why can this not be implemented now? AFAICT it does not depend on challenge response or any missing functionality.

File: examples/aggregator/module/rust/src/lib.rs
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs:91]: Is this accurate? AFAICT the Router node has public label.

File: examples/aggregator/module/rust/src/lib.rs
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs:120]: This one will have to change too, right? It needs to have the wasm hash label, and will declassify when writing out to the backend.

File: examples/aggregator/module/rust/src/lib.rs
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs:128]: Does handler == aggregator == Aggregator? If so, could you make the messages and comments more consistent?

File: examples/aggregator/module/rust/src/lib.rs
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs:173]: I think this is the backend server address, right? If so, could you name it that way?

**Comment by ipetr0v** [comment]
Currently blocked because `Router` needs to be initialized, but it can only have a gRPC channel receiver.

Originated from: https://github.com/project-oak/oak/pull/1653#discussion_r515148501

File: examples/aggregator/module/rust/src/lib.rs
Comment (tiziano88) [reply examples/aggregator/module/rust/src/lib.rs:93]: Just checking: do you mean this corresponds to the TODO above, right?

File: examples/aggregator/config.toml
Comment (tiziano88) [inline examples/aggregator/config.toml:3]: I guess this was the only way of avoiding the infinite regress of the hash including itself as part of the hash calculation? :) Apart from splitting the aggregator into a separate wasm module (which BTW I guess would be the correct way of doing this, since currently this hash includes all the nodes, including the public ones, so if you change a single string in the router node, for instance, this hash will also change).

**Comment by tiziano88** [comment]
> Currently blocked because `Router` needs to be initialized, but it can only have a gRPC channel receiver.
> 
> Originated from: [#1653 (comment)](https://github.com/project-oak/oak/pull/1653#discussion_r515148501)

We should probably implement the Init pattern that @wildarch suggested (can be a special case for now, like Either) to unblock this?

**Comment by tiziano88** [comment]
I just hacked this together: https://github.com/project-oak/oak/pull/1659/files I think it would solve this case (and in general, the initialization pattern), do you think that is sufficient @ipetr0v ? also cc @wildarch .

File: examples/aggregator/config.toml
Comment (ipetr0v) [reply examples/aggregator/config.toml:3]: I think we had this pattern in TIR (separate modules for nodes) but then decided to merge them together.
Not sure that we need separate modules in this example, since this example is very simple right now.

Comment (ipetr0v) [reply examples/aggregator/config.toml:3]: And also it would require to make separate modules in all examples, where we want to check incoming modules

Comment (tiziano88) [reply examples/aggregator/config.toml:3]: It would only be necessary for nodes that perform declassification based on wasm hash or signatures. For others, there is no point in splitting them.

**Comment by ipetr0v** [comment]
Unblocked by https://github.com/project-oak/oak/pull/1675

File: examples/aggregator/proto/handler_init.proto
Comment (ipetr0v) [reply examples/aggregator/proto/handler_init.proto:19]: Done

File: examples/aggregator/config.toml
Comment (ipetr0v) [reply examples/aggregator/config.toml:3]: Should we add a function that returns hash and public key of a module, for these purposes?

Comment (ipetr0v) [reply examples/aggregator/config.toml:3]: And we probably wouldn't need to split modules (not sure yet)

Comment (ipetr0v) [reply examples/aggregator/config.toml]: Will divide modules as a separate PR

File: examples/aggregator/module/rust/tests/integration_test.rs
Comment (conradgrobler) [inline examples/aggregator/module/rust/tests/integration_test.rs:55]: The hash is repreated in many places. Perhaps add documentation somewhere to explain how to get the new hash and all the different places where the hash needs to be updated if the aggregator example needs to be rebuilt.

File: examples/aggregator/proto/BUILD
Comment (conradgrobler) [inline examples/aggregator/proto/BUILD:48]: Is this required? The proto files are only used by the rust code on the server side, not by any clients.

**Review by conradgrobler — COMMENTED** [review]
The CI is failing. It seems that the request receiver on the invocation has an empty label for some reason.

File: examples/aggregator/module/rust/tests/integration_test.rs
Comment (ipetr0v) [reply examples/aggregator/module/rust/tests/integration_test.rs]: Done

**Comment by ipetr0v** [comment]
CI is failing because we need labeled channels in `oak_tests`
Done in https://github.com/project-oak/oak/pull/1706 (will rebase this PR once it's merged)

**Review by conradgrobler — APPROVED** [review]

File: examples/aggregator/module/rust/src/lib.rs
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs:74]: Am I reading this correctly that it is sending the entire configmap to the router node? That seems excessive, also the config in general may contain other details that are irrelevant for the router node. AFAICT the only reason for passing it is in order to create the gRPC client node (which I think should be created here) and in order to pass the wasm module hash (which I think should be hardcoded here, though that's not possible yet because the app is still in a single wasm module for now). Maybe leave a TODO for the latter for now? Though I think the former could be fixed now already.

**Review by tiziano88 — APPROVED** [review]

File: examples/aggregator/module/rust/src/router.rs
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/router.rs:80]: I think, specifically for this it could be a wrapper around Channel, that checks privileges (probably does the same thing as Channel in Runtime), but does this in SDK.
Since having SDK (non-Runtime) errors was the only reason behind having this piece of code.

File: examples/aggregator/module/rust/src/router.rs
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/router.rs:100]: Handler is also created inside `Router`. Will probably create all nodes in the Main node.
But this will leave `Router` only the check of incoming labels (which later on, I think, should be a part of SDK).

File: examples/aggregator/module/rust/src/lib.rs
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/lib.rs:74]: Adding a TODO(https://github.com/project-oak/oak/issues/1731)

File: examples/aggregator/client/android/cpp/client.cc
Comment (ipetr0v) [reply examples/aggregator/client/android/cpp/client.cc]: Done

File: examples/aggregator/config.toml
Comment (ipetr0v) [reply examples/aggregator/config.toml]: Correct, I will remove this as part of https://github.com/project-oak/oak/issues/1731

**Pull request merged**
